feat(extract): add concatenation reduction

kopecs opened the pull request:
This PR also cleans up the existing code a bit.

PR checklist:

- [x] Documentation is up-to-date (to be updated on merge)
- [x] `changelog.d/<issue>.<type>` is a file with the _what_, _why_, and _how_ of the change.
  - \<issue> is `pa-312` (Linear ticket), `gh-1234` (GitHub issue), or `new-gizmo` (unique semantic name)
  - \<type> is `added`, `changed`, `fixed`, or `infra`.
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see the [Contribution guidelines](https://semgrep.dev/docs/contributing/how-to-contribute/)!

Test plan: `make test`

Commit:
Add a concatentation reduction mode for extraction

Refactor and cleanup the extraction code as well.

Commit:
Add extract reduction to parser and schema

Commit:
Add changelog entry

mmcqd commented on semgrep-core/src/extracting/Extract.ml:
```diff
@@ -39,82 +39,269 @@ let ( let* ) = Option.bind
 (*****************************************************************************)
 
+(* A type for nonempty lists *)
+type 'a nonempty = Nonempty of 'a * 'a list
```
Maybe move to Common?

mmcqd commented on semgrep-core/src/extracting/Extract.ml:
```diff
@@ -39,82 +39,269 @@ let ( let* ) = Option.bind
   let rule_pairs = Common.map (fun r -> (fst r.Rule.id, r)) rules in
   Common.hash_of_list rule_pairs
 
+(** Collects a list into a list of equivalence classes (themselves nonempty
```
Getting 150 flashbacks

kopecs replied:
That's a lot of flashbacks to get.

mmcqd commented on semgrep-core/src/extracting/Extract.ml:
```diff
@@ -39,82 +39,269 @@ let ( let* ) = Option.bind
 
-(** This is the main function which performs extraction of the matches
-   generated by extract mode rules.
+let extract_collect_and_comb erule_table xtarget rule_ids matches =
```
This function is pretty brutal, but you've done a nice job making it elegant. I might add a comment above each `|> ...` summarizing what that step is doing (at least the ones that are more than like 2 lines).

mmcqd reviewed (approved):
This needs a test plan, ideally a test with a concat rule.

Commit:
Improve source-level docs

Commit:
Don't use the target path as base for tmpfile name

This can cause issues if the target path includes directory separators,
and is really just unnecessary.

Commit:
Move `Extract` -> `Match_extract_mode`

This avoids a circular dependency between engine and extracting (by
eliminating extracting). Namely, extracting relied on engine for
Match_rules.check and engine would rely on extracting for testing
purposes.

Commit:
Add test infrastructure for extract mode

Commit:
Add test for extract mode

Commit:
squash! Add test for extract mode

Commit:
fixup! squash! Add test for extract mode

emjin commented on semgrep-core/src/core/Rule.ml:
```diff
@@ -196,12 +196,15 @@ type taint_spec = {
+    - either treat them as separate files; or
+    - concatentate them together
+*)
+type extract_reduction = Separate | Concat [@@deriving show]
```
Extract_method? Extract_treatment_of_ranges? (I don't have good ideas I just don't love the current name, as I read more of this file though I think it's less important)

kopecs replied:
I think the bigger question would be for the user facing yaml key which corresponds to this. I don't think the internal name of this type is significant.

kopecs replied:
Do you want to hash this out now, or just merge this with a disposition to revisit naming there later?

emjin replied:
Merge it

emjin commented on semgrep-core/src/engine/Match_extract_mode.ml:
```diff
@@ -0,0 +1,363 @@
+(* Main logic *)
+(*****************************************************************************)
+
+let extract_collect_and_comb erule_table xtarget rule_ids matches =
```
Since you call it Concat in your type, it would be helpful to have "concat" in this function name

emjin commented on semgrep-core/src/engine/Match_extract_mode.ml:
```diff
@@ -0,0 +1,363 @@
+            | None -> None))
+  (* Factor out rule *)
+  |> List.filter_map (function
+       | [] -> None
```
Just a question: why not factor out the rule earlier, when you still have nonempty lists?

kopecs replied:
We don't actually have the rule itself until we look it up in the table in the step immediately prior to this; we just have the ruleid on each match before then.

emjin commented on semgrep-core/src/engine/Match_extract_mode.ml:
```diff
@@ -0,0 +1,363 @@
+                    start_col = snippet_trailing;
+                  },
+                  contents,
+                  (* Map results by chaining functions together, successively
```
Would be helpful to note more clearly that this is how you're mapping back the results to the original. Also helpful to add a note that matches that cross boundaries are reported in the first snippet

Commit:
Address Emma's feedback

aryx commented on semgrep-core/src/engine/Match_extract_mode.ml:
```diff
@@ -0,0 +1,374 @@
+         |> List.rev
+         (* Read the extracted text from the source file *)
+         |> Common.map (fun { start_pos; start_line; start_col; end_pos } ->
+                let source_file = open_in_bin xtarget.Xtarget.file in
```
you forgot to close_in here, leading to "too many opened files errors".
(I've fixed it today).

kopecs replied:
Whoops, sorry!

aryx commented on semgrep-core/src/engine/Test_engine.ml:
```diff
@@ -160,25 +184,70 @@ let make_tests ?(unit_testing = false) xs =
+                         xtarget
+                     in
+                     Hashtbl.find_opt extract_result_map file
+                     |> Option.fold ~some:(fun f -> f res) ~none:res)
```
Those Option.fold are ugly IMHO. They don't save you anything compared to just
match xxx with
| Some f -> f res
| None -> res
